### ISSUE TITLE
Tools without a container should not be legal CWL

### ISSUE THREAD

**Issue by kaushik-work** [common-workflow-language/common-workflow-language#782]
I was surprised to find that we allow tools with a command line but no docker/container requirement to validate and `cwltool` runs these tools. This is extremely bad practice from a reproducibility and portability point. 

We should enforce good practice by not allowing non-containered tools.

**Comment by mr-c** [comment]
Hey @kaushik-work ,

I think a `--strict` mode for the CWL reference runner would be a good way to implement this.

Not all CWL users use software containers, some use locally installed software, software installed from package manager (conda/bioconda/Debian/Guix), etc.

The philosophy has always been to enable reproducibility but not to demand it. Besides, even a `Dockerfile` or `dockerPull: alpine:latest` isn't reproducible :-)

**Comment by kaushik-work** [comment]
> Not all CWL users use software containers, some use locally installed software, software installed from package manager (conda/bioconda/Debian/Guix), etc.

> The philosophy has always been to enable reproducibility but not to demand it.

> Besides, even a Dockerfile or dockerPull: alpine:latest isn't reproducible

The problem with all this is that workflows get released - sometimes as part of grant deliverables - that have hidden assumptions about software versions and running platform environment. With us striving for reusability and modularity this means workflows can be unknowingly "infected" with non-reproducible and non-portable components. This will turn out to be a time-bomb and a long term disaster.

Does `--strict` now exist? I think default validation should flag anything that will lead to non-portable, non-reproducible workflows by default.

**Comment by psafont** [comment]
> The problem with all this is that workflows get released - sometimes as part of grant deliverables - that have hidden assumptions about software versions and running platform environment.

Container references may be of some help, but they do not solve this issue.

Take for example a reference to a container hosted in a private repository, or the repository may even be moved to another address or even disappear, and then we have the problems with container tagging which @mr-c mentions.

**Comment by mr-c** [comment]
> The problem with all this is that workflows get released - sometimes as part of grant deliverables - that have hidden assumptions about software versions and running platform environment. With us striving for reusability and modularity this means workflows can be unknowingly "infected" with non-reproducible and non-portable components. This will turn out to be a time-bomb and a long term disaster.

This is one of the things Research Objects & CWLProv are aiming to solve. For example: `dockerPull: alpine:latest` we'll be recording the exact container used, not just the has. We may even include the container image itself in the RO, in an "offline" archival mode.

> Does `--strict` now exist? I think default validation should flag anything that will lead to non-portable, non-reproducible workflows by default.

Not yet. A PR to issue a warning would be welcome. We currently do this when running under MS Windows.

**Comment by stain** [comment]
I think it should still be legal, as containers is just one of the way to distribute and activate software packages there's also BioConda and on HPC [modules](http://modules.sourceforge.net/) is common. 

There's also the philosophy that CWL should allow you to do a gradual move from not-so-reproducible local shell script to a fully interoperable container-based workflow. So we should not ban execution of slightly "scruffy" workflows. 

That said, we should change our CWL spec to RECOMMEND a `SoftwareRequirement` so that engines can issue warnings if it is not included.

**Comment by kaushik-work** [comment]
> Take for example a reference to a container hosted in a private repository, or the repository may even be moved to another address or even disappear, 

Yes, this is a deep problem. For published workflows, the docker images used should be deposited in durable data repositories, like we are pushing for raw data. 

> and then we have the problems with container tagging 

Yes, absolutely: So we should push for using identifying images by digest.

> This is one of the things Research Objects & CWLProv are aiming to solve.

I have a strong opinion that we should not be tacking on complexity when it is not needed. By enforcing discipline at the heart of workflow development we support reproducibility in science.

**Comment by mr-c** [comment]
@kaushik-work You are very welcome to be more strict in Rabix Composer, or to write documentation that always calls `cwltool --strict`, etc..

But support for non-container users has always been a priority for this project, though with the arrival of `udocker` that is less urgent than it used to be.

> So we should push for using identifying images by digest.
That's not wrong, but your users won't like it. Very human unfriendly.

about RO & CWLProv:

Think of a CWL document as a (slightly abstract) plan: it doesn't specify which data, or where to run, or the exact CPU type, etc..

When that plan is realized into a concrete execution, that is what we communicate with RO/CWLProv. They serve different purposes and different use cases.

**Comment by kaushik-work** [comment]
@stain you bring up good points!

> I think it should still be legal, as containers is just one of the way to distribute and activate software packages there's also BioConda and on HPC modules is common.

Agreed: I'm not for prescribing a particular technology - but it must be a technology, like containerization, that gives strong guarantees of what environment the tool will end up running in.

> There's also the philosophy that CWL should allow you to do a gradual move from not-so-reproducible local shell script to a fully interoperable container-based workflow. So we should not ban execution of slightly "scruffy" workflows.

In my opinion there is no middle ground here :D. Either we stick with Bash scripts or we do CWL right, which means no sloppiness :D. Once we create a crack, people will use push things through it in the name of "We'll fix it later". If you are at that stage, use bash until you are ready, then describe things in CWL properly.

> That said, we should change our CWL spec to RECOMMEND a SoftwareRequirement so that engines can issue warnings if it is not included.

My last fall back position, from which I will not budge :D is that engines by default will complain mercilessly and loudly when you engage in bad practices. They are even allowed to not run your workflow. And yes, we need a best practices document. I am delighted to contribute to it, though my experiences are probably limited.

**Comment by FarahZKhan** [comment]
I agree with Stian here about a gradual move from conventional to reproducible/interoperable workflows. I think as a standard, CWL can "highly encourage"/"recommend" and "support" the best practices and then provide constructs/implementations as well as workflow examples to highlight the impact of these best practices. 
If we dont allow the local installations and make containerization a MUST, it will add an additional level of complexity for people to resolve, hence might not address requirements of all types of audience.

**Comment by kaushik-work** [comment]
@mr-c, @stain @FarahZKhan  I think I understand where you and others are coming from - you want to make it easy for people to transform a workflow into CWL by skipping the tool packaging step. If it were easier to CWL-ize a tool, more people would use it.

My philosophy is that workflow systems have been around since the '70s. People have used tons of workflow systems on their local environment, often self developed. There is no dearth of options to get something up and running on your own, controlled environment.

What I see CWL doing is to prevent these non-reproducible, non-portable workflows and **enforce** good practices. Make people strongly aware that while what they are doing works for them right now, it probably will fail on a foreign environment.

If they don't care about foreign environments, why are they using CWL? A bash script, or a Python script or some other (often bespoke) workflow framework is just as good. :)

**Comment by jpellman** [comment]
I think one point that's being lost in this conversation is that containerization is not necessarily equivalent to packaging and versioning software or documenting a computational environment.  Containerization, as it stands today, is basically a virtual environment for system libraries (i.e., Python:venv, C/Linux:Docker), with fine-grained controls for how resources are used that allow you to oversubscribe resources on a beefy bare metal server (or a VM in some cases).

In my opinion, packaging and versioning is mostly a different problem domain that is better solved by tools like yum/apt/pip/etc.  Yes, you can package an app with multiple dependencies by containerizing it, but in doing so you've effectively eliminated the _how_  and _why_ of your in silica setup by converting it into a black box.  You're also making the ability to remix your setup that much harder. 

In contrast, if you package multiple versions of a piece of software into an RPM repository, it becomes trivial to fetch the exact version of the software and all of its required dependencies.  You can then document a particular setup using a configuration management tool like [Ansible](https://www.ansible.com/), which uses a declarative syntax (a YAML file, just like CWL) that allows you to name the exact versions you used for your analysis.  If a future researcher wants to use a modified version of your setup with only one different version of an app, for instance, all that needs to change is a single field in the config management script.

**Comment by jpellman** [comment]
To add to my earlier comment, it's worth taking a look at the article [Why Docker is not the answer to reproducible research, and why Nix may be](https://github.com/WeAreWizards/blog/blob/d5611b00e238115a6904f5f9e9d8772161e46c8e/content/articles/reproducible-research-or-why-docker-is-not-the-answer.md), which I believe provides compelling arguments in favor of the use of the Nix package manager (which is light-years ahead of yum/apt/pip in its sophistication) for producing reproducible in silica setups for scientific computing over relying upon Docker containers alone.

See too:

* [Reproducible Environments With Nix](https://www.software.ac.uk/blog/2017-10-05-reproducible-environments-nix)
* [A truly reproducible scientific paper](https://mozillafoundation.github.io/2017-fellows-sf/re-papers/)

**Comment by mr-c** [comment]
Everyone, thank you for the conversation. If someone has a specific proposal then we should make a specific issue & pull request to implement that proposal.

@jpellman: Hello! Thank you for joining the conversation. 

I and many others in the CWL project agree that containerization != packaging.

Are you familiar with https://www.commonwl.org/v1.0/CommandLineTool.html#SoftwareRequirement & https://github.com/common-workflow-language/cwltool#leveraging-softwarerequirements-beta ?

It is likely that someone could write a dependency resolver configuration for Nix/Guix. Such a configuration file & documentation would be a welcome contribution to the CWL reference runner (`cwltool`). Likewise we need to extend the `cwlprov` work to capture the (Nix, Conda, etc..) executable environment when a `SoftwareRequirement` is used.

For CWL v1.x (and likely v2.x) the serialization format for command line analysis tools is the docker container image format. That docker container might be derived from Nix, Conda, Debian, or a regular (and fragile) Dockerfile. Likewise that docker container might be executed using the Docker.com daemon, runc, Singularity, udocker, or any other compatible container runtime.

**Comment by pjotrp** [comment]
I reran the Guix-based [example]( https://hpc.guix.info/blog/2019/01/creating-a-reproducible-workflow-with-cwl/) with a recent CWL in GNU Guix. See https://github.com/pjotrp/CWL-workflows. 

It uses --no-container and finds the tools in the *running* container (if they are in the path - i.e. CWL is running inside a container with all tools in the path). The one exception where I need something from the environment is looking for a jar which is a directory specified as 

    GUIX_ENVIRONMENT=/gnu/store/v2jys382g6j5b7lsxzh8v4vfhd414nhz-profile

Currently I hard code that in the tool (see [definition#L58](https://github.com/pjotrp/CWL-workflows/blob/154ab01371c5176d3a3fc5cf3ce33ad9641021f5/Tools/trimmomaticPE.cwl#L58)). This is not ideal, obviously. Would it be an idea to have a variable guix.profile which can expand with something like 

```sh
valueFrom: $(guix.profile)/lib/share/jar/trimmomatic.jar
```

or is there another elegant solution? I am happy to come up with something generic for GNU Guix.

**Comment by mr-c** [comment]
@pjotrp For Java, when there isn't a normally available script to run, I recommend that CWL CommandLineTool descriptions _do not_ refer to the jar path (even if using `DockerRequrement` but instead refer to the main entry point on and rely on `CLASSPATH` being set correctly by the container or the workflow engine.

`baseCommand: [java, org.usadellab.trimmomatic.Trimmomatic]`

For GUIX, it should set `CLASSPATH=/gnu/store/v2jys382g6j5b7lsxzh8v4vfhd414nhz-profile/lib/share/jar/trimmomatic.jar`

Or GUIX should put a script named `trimmomatic` on the PATH, like the biocontainer does.

**Comment by pjotrp** [comment]
> Or GUIX should put a script named `trimmomatic` on the PATH, like the biocontainer does.

That is a good idea. It means we don't need special treatment of Guix ;). I do need to find other files too, such as     

    valueFrom: 'ILLUMINACLIP:/gnu/store/v2jys382g6j5b7lsxzh8v4vfhd414nhz-profile/lib/share/jar/adapters/TruSeq2-PE.fa:2:40:15'

which gets installed with the jar. I suppose the script could pick that up, but it would be nice to parametrize. What is the best way to pick up an environment variable?

**Comment by pjotrp** [comment]
If I can't pick up the setting in the environment, I suppose I can simply generate the YAML before kicking off the workflow. I think, in time, I can create CWL workflows from GNU Guix directly. 

On a separate note: I don't think CWL should make tools without containers illegal because we can make the environment with tools reproducible by embedding CWL in Guix. This issue can therefore be closed in my opinion. One of the features of CWL is that it is non-opiniated about tools.
